Introduction of namespaced features (see #1286)

djc opened the pull request:
I think this basically covers all of the plans from #1286, although it still needs a bunch of tests and documentation updates. Submitting this PR to get some early feedback on the direction.

alexcrichton commented:
Thanks! I'm have a bit of a hard time following this though, mind adding some comments on the various `bail!` statements in the updated feature resolution loop?

Additionally, to recap, the idea here is that packages with this flag *only* activate optional dependencies through `crate:foo`, right?

djc commented:
@alexcrichton I've added a whole bunch of comments.

To recap: the idea here is that features live in a separate namespace from dependencies. As such, when specifying dependencies in a feature value list (in the features table), you have to prefix dependencies with `crate:`. However, even after this PR, you do not have to specify a feature explicitly if you just have an optional dependency: for optional dependencies where no feature is explicitly defined, an implicit feature will be created -- this should help avoid boilerplate. Then again, if you do specify a feature of the same name as an optional dependency, it *must* include the `crate:feature` dependency.

Hope that makes sense.

alexcrichton commented on src/cargo/core/summary.rs:
```diff
@@ -131,55 +137,135 @@ impl PartialEq for Summary {
                     dep
                 ),
-                (&CrateFeature(_, _), Some(_)) | (&Feature(_), _) => {}
+                (&CrateFeature(_, _), _, _) => {}
```
Could this explicitly list `(&CrateFeature(_, _), true, _)` for exhaustiveness?

djc replied:
Fixed.

bors commented:
:umbrella: The latest upstream changes (presumably #5374) made this pull request unmergeable. Please resolve the merge conflicts.

djc commented:
Added a number of test cases that also did a good job flushing a number of logic issues out of the implementation. What is required in terms of documentation?

One thing to consider: should `namespaced-features` default to `true` for `Edition2018`?

bors commented:
:umbrella: The latest upstream changes (presumably #5335) made this pull request unmergeable. Please resolve the merge conflicts.

djc commented:
I've rebased on latest master and added some documentation to the unstable section.

djc commented on src/cargo/core/summary.rs:
```diff
@@ -131,55 +137,168 @@ impl PartialEq for Summary {
             values.push(val);
         }
+
+        if !dependency_found {
```
I added a whole bunch of comments, does that help? It spans across the for-loop (which includes the match) as a performance optimization: this way, we don't have to loop over the feature values twice. That's entirely premature (in the sense of not having done any performance measurements), so if the comments don't help enough, I could instead rewrite to check the `values` list after the loop.

alexcrichton commented on src/cargo/util/toml/mod.rs:
```diff
@@ -763,6 +764,7 @@ impl TomlManifest {
             deps,
             me.features.clone().unwrap_or_else(BTreeMap::new),
             project.links.clone(),
+            project.namespaced_features.unwrap_or(false),
```
Ah I just remembered, can this be gated behind a feature as well? (setting this key at all)

djc replied:
Sure - not sure what kind of gate you had in mind?

alexcrichton replied:
Oh something like `cargo-features = ["namespaced-features"]` in `Cargo.toml`, you can see some other examples of this via `src/cargo/core/features.rs` I think

alexcrichton replied:
For now yeah we'll want that (the Cargo feature basically unlocks the `namespaced-features` config value).

Long term we may not have both but for the near-term I think we'll want that.

alexcrichton commented:
Ok thanks for the new comments, they look great! I think this is good to go with a feature gate and fixed tests!

djc commented:
Huh, those test failures got introduced by merging a conflict wrong. Fixed now, and rebased on current master. Just let me know what kind of feature gate you want and this can (finally!) be merged.

bors commented:
:umbrella: The latest upstream changes (presumably #5384) made this pull request unmergeable. Please resolve the merge conflicts.

djc commented:
Rebased on master and added the feature gate. Hope we can merge this now!

alexcrichton commented:
Hm looks like CI may be failing?

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 0b6f420 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 0b6f420670cef2bbfc9ea9705e46d2479d308a11 -->

bors commented:
:hourglass: Testing commit 0b6f420670cef2bbfc9ea9705e46d2479d308a11 with merge 4963d884a4f39371b31cccd4db28297cdbeb24b5...
